cpp/src/arrow/util/decimal.cc

```diff
@@ -41,11 +42,13 @@ Decimal128::Decimal128(const std::string& str) : Decimal128() {
 }
 
 Decimal128::Decimal128(const uint8_t* bytes)
-    : Decimal128(reinterpret_cast<const int64_t*>(bytes)[0],
-                 reinterpret_cast<const uint64_t*>(bytes)[1]) {}
+    : Decimal128(BitUtil::FromLittleEndian(reinterpret_cast<const int64_t*>(bytes)[1]),
+                 BitUtil::FromLittleEndian(reinterpret_cast<const uint64_t*>(bytes)[0])) {
+}
 
 std::array<uint8_t, 16> Decimal128::ToBytes() const {
-  const uint64_t raw[] = {static_cast<uint64_t>(high_bits_), low_bits_};
+  const uint64_t raw[] = {BitUtil::ToLittleEndian(low_bits_),
+                          BitUtil::ToLittleEndian(static_cast<uint64_t>(high_bits_))};
```

Comment:
`reinterpret_cast` here -- can this change the bit pattern?

Reply:
`static_cast` from one builtin integer type to another of the same byte size won't change the underlying bits.

The compiler will also fail to type check this because `reinterpret_cast` from one integer to another isn't allowed by the standard: http://en.cppreference.com/w/cpp/language/reinterpret_cast.

Reply:
Just to clarify where this behavior is defined in the standard: http://en.cppreference.com/w/cpp/language/implicit_conversion#Integral_conversions

The specific language is this:
> If the destination type is unsigned, the resulting value is the smallest unsigned value equal to the source value modulo 2 ** n
 where n is the number of bits used to represent the destination type.
That is, depending on whether the destination type is wider or narrower, signed integers are sign-extended [footnote 1] or truncated and unsigned integers are zero-extended or truncated respectively.

Here's the footnote :)
> This only applies if the arithmetic is two's complement which is only required for the exact-width integer types. Note, however, that at the moment all platforms with a C++ compiler use two's complement arithmetic

We're using exact-width integer types here, so all behavior here is well-defined and within the standard.

Reply:
Gotcha. there is always `*reinterpret_cast<T*>(&val)` but this is helpful to know